src/workerd/util/weak-refs.h

```diff
@@ -51,4 +51,40 @@ class AtomicWeakRef final: public kj::AtomicRefcounted {
   friend T;
 };
 
+// A WeakRef is a weak reference to a thing.
+// TODO(cleanup): It would eventually be nice to replace kj::Own<WeakRef<T>> with a
+// kj::WeakOwn<T> type with the same basic characteristics.
+template <typename T>
+class WeakRef final: public kj::Refcounted {
+public:
+  inline WeakRef(kj::Badge<T>, T& thing) : maybeThing(thing) {}
+  KJ_DISALLOW_COPY_AND_MOVE(WeakRef);
+
+  // Run the functor and return true if the context is alive, otherwise return false. Note that
+  // since the `IoContext` might not be alive for any async continuation, we do not provide
+  // a `kj::Maybe<IoContext&> tryGet()` function. You are expected to invoke this function
+  // again in the next continuation to re-check if the `IoContext` is still around.
+  template<typename F>
+  inline bool runIfAlive(F&& f) const {
+    KJ_IF_SOME(thing, maybeThing) {
+      kj::fwd<F>(f)(thing);
+      return true;
+    }
+
+    return false;
```

Comment:
So this is the same as:
```
  return maybeThing.map(kj::fwd<F>(f)) != kj::none;
```
For consistency with `tryGet` it would be better called `tryRun`. It's also not clear to me that this better to have this than just calling `map` directly...

Reply:
Essentially just copied this directly. I don't want to rename this since there is existing code in the internal repo that would be impacted by a outright rename. If we were going to rename it I would first introduce an alias, switch the internal code over to using the alias, then remove the original name once the internal repo has been updated. That said, Not really seeing much reason to rename it at this point.

Reply:
And the impls there are not quite the same. Passing `f` into the `map()` function would make use of its return value, if any, and the original ignores it. It's a slight difference but could have a definite impact on the return value.